### PR TITLE
feat: add docker support

### LINKED ISSUE #3896: Dockerization Proposal for heyxyz/hey
### Summary

**Proposal:**

I propose the Dockerization of the **heyxyz/hey** open-source project to enhance its accessibility, reproducibility, and deployment efficiency. Dockerization involves creating a Docker image and Docker Compose configurations that encapsulate the project's dependencies and environment settings.

**Benefits:**

**Simplified Setup:** Dockerization will streamline the setup process for newcomers and eliminate issues related to different development environments.

**Reproducibility:** With a Docker image, users can easily reproduce the project's environment, ensuring consistency across deployments.

**Portability:** Docker containers can run on various platforms, including local development machines, cloud servers, and CI/CD pipelines.

**Isolation:** Docker containers provide isolation, preventing conflicts with system-level dependencies.

**Discussion:** Initiate a discussion with the project maintainers and community members to finalize the Dockerization plan.

**Execution:** Execute the Dockerization process, following the defined plan and guidelines.

**Testing:** Thoroughly test the Docker image to ensure it functions as expected.

**Documentation Update:** Update the project's documentation to include Docker-related instructions.

**Announcement:** Announce the availability of the Dockerized version to the community and provide guidance on its usage.

### PR DESCRIPTION
**Pull request by ezhil56x**
## What does this PR do?
- This PR is to implement dockerization to heyxyz/hey

## Related issues

Fixes #3896
/claim #3896  

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (non-breaking small changes to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Explanation of the changes
- Implemented dockerization to heyxyz
- Can build docker images of `api`, `web`, and `og`
- Setup workflow that gets triggered when there is a push on the repo
- Workflow builds and pushes the images to dockerhub

### CODE REVIEW

**Comment by bigint** [comment]
@ezhil56x can you make a short video running this on docker 🙇🏼 ?

**Comment by ezhil56x** [comment]
@bigint, Here is the working demo. Most time was for building images. Skip to last few minutes to see deployment

https://github.com/heyxyz/hey/assets/103899034/071bb25f-34f4-42ef-87c7-115e4c0dd40c

**Comment by bigint** [comment]
manual env copy are hard with headless machines like railway 🙈

**Comment by ezhil56x** [comment]
> manual env copy are hard with headless machines like railway 🙈

Only when we pass env as file in compose, it's working. Can I add a script than?

**Comment by bigint** [comment]
for compose its fine, but while deploying docker to prod? is it possible to pass env vars?

**Comment by ezhil56x** [comment]
> for compose its fine, but while deploying docker to prod? is it possible to pass env vars?

Yeah it is possible to pass through `docker run -e VARIABLE ...`

**Comment by ezhil56x** [comment]
@bigint Any update? 🙈

**Comment by bigint** [comment]
Can you deploy OG app to dockerhub and deploy in railway?

- We are missing devcontainers
- We are missing deployment workflows

**Comment by ezhil56x** [comment]
> Can you deploy OG app to dockerhub and deploy in railway?
> 
> * We are missing devcontainers
> * We are missing deployment workflows

As discussed before having re

> Can you deploy OG app to dockerhub and deploy in railway?
> 
> * We are missing devcontainers
> * We are missing deployment workflows

As discussed before, hot reload uses only the node runtime instead of the run time that's on the host

Do you want me to deploy only the og app that is under `apps/og` for now?

**Comment by bigint** [comment]
Yes you can deploy, so if we using only for deployment why we need compose file?

**Comment by ezhil56x** [comment]
> Yes you can deploy, so if we using only for deployment why we need compose file?

Compose file is easy to use than `docker run ... ` and passing the env in cli

**Comment by bigint** [comment]
Who will be the users gonna use docker compose? devs / devops?

**Comment by ezhil56x** [comment]
> Who will be the users gonna use docker compose? devs / devops?

Devops

**Comment by bigint** [comment]
Do they gonna use docker compose? im not sure it will be helpful if we are using railway 🙇🏼 cuz each service will be deployed on different servers

**Comment by ezhil56x** [comment]
> Do they gonna use docker compose? im not sure it will be helpful if we are using railway 🙇🏼 cuz each service will be deployed on different servers

That can be removed if not needed. I have developed on Railway. Can you check you Dm on WA

File: .github/workflows/dockerhub.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Build and push web image
```
Comment (bigint) [inline .github/workflows/dockerhub.yml:27]:
```suggestion
      - name: Build and push Web image
```

File: .github/workflows/dockerhub.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          file: ./apps/web/Dockerfile
+          push: true
+          tags: ${{ secrets.DOCKERHUB_USERNAME }}/web:latest
+
+      - name: Build and push api image
```
Comment (bigint) [inline .github/workflows/dockerhub.yml:35]:
```suggestion
      - name: Build and push API image
```

File: .github/workflows/dockerhub.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          file: ./apps/api/Dockerfile
+          push: true
+          tags: ${{ secrets.DOCKERHUB_USERNAME }}/api:latest
+
+      - name: Build and push og image
```
Comment (bigint) [inline .github/workflows/dockerhub.yml:43]:
```suggestion
      - name: Build and push OG image
```

File: .github/workflows/dockerhub.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+name: Build and Push to Docker Hub
+
+on:
+  push:
+    branches:
+      - main
```
Comment (bigint) [inline .github/workflows/dockerhub.yml:6]:
```suggestion
on:
  push:
    branches: [main]
```

**Review by bigint — CHANGES_REQUESTED** [review]

**Review by bigint — APPROVED** [review]
**LGTM** :shipit: 

LGTM

**Pull request merged**